torch/_inductor/codegen/cpp_gemm_template.py

```diff
@@ -851,12 +853,24 @@ def maybe_to_dense(inputs, layout_or_out):
         def normalize_shapes(inputs, layout_or_out):
             new_inputs = list(inputs)
             if not is_mkldnn_wgt and isinstance(new_inputs[1], torch.Tensor):
-                if view_size[0].is_symbol:
+                if has_free_symbols(view_size):
                     # If batch size B is dynamic, we need to infer the batch size from the input
-                    assert all(not dim.is_symbol for dim in view_size[1:])
+                    assert not has_free_symbols(view_size[1:])
                     size = torch.tensor(new_inputs[1].size()).prod()
                     fixed_size = torch.tensor(view_size[1:], dtype=torch.int).prod()
+                    eq = sympy.Eq(view_size[0], int(size // fixed_size))
                     view_size[0] = (size // fixed_size).item()
+
+                    # solve for any free symbols, since they might exist in stride (in HF XLNetLMHeadModel dynamic case)
+                    free_symbols_prod = math.prod(eq.free_symbols)
+                    free_symbols_equivalent = sympy.solve(eq, free_symbols_prod)[0]
+                    if has_free_symbols(view_stride):
+                        for i, sym in enumerate(view_stride):
+                            if has_free_symbols((sym,)):
+                                view_stride[i] = sym.xreplace(
+                                    {free_symbols_prod: free_symbols_equivalent}
+                                )
```

Comment:
Can you give an example of this is doing?   Do we need to handle the case where the equation fails to solve?

Reply:
The following BMM op exists in the HF XLNetLMHeadModel:
```python 
bs, Mdim = 8, 512
u = torch.randn(bs, 16, Mdim, Mdim)
v = torch.randn(1, bs * Mdim, 2 * Mdim)
torch._dynamo.mark_dynamic(u, 0)
torch.bmm(
    u.reshape(-1, Mdim, Mdim),
    v.reshape(Mdim, bs * 16, 64).permute(1, 0, 2)
)
```
This gives a layout of `[64, 1024 * s0, 1]` for the stride of the second matrix.

The code above is intended to resolve any symbols in the stride. There are guards in `use_cpp_gemm_template` which should prevent dynamic shapes in the `n` and `k` dims, so there should be at most one expression in view_size and view_stride with symbols. Since the symbol in view_size can always be inferred (because the tensor is realized), we can also solve the symbol in view_stride.

Essentially, we will be able to solve this equation because there will only be one dynamic expr, in the batch dim.

Reply:
Take a further look of these code: https://github.com/pytorch/pytorch/blob/c83ef7aa9feccf2f44c06ce684ee64212ee0c253/torch/_inductor/codegen/cpp_gemm_template.py#L857C1-L872C34

For BMM case, there might be 2 cases

- Case 1: the `new_inputs[1]` is not a constant, then these code should only be hit once in the benchmark phase from `preprocessor`. Then how about we directly use `new_view_size = V.graph.sizevars.size_hints(view_size)` to instance the values. I think it will simplify these code.
- Case 2: the `new_inputs[1]` is a constant. Then the `view_size` of `new_inputs[1]` shouldn't include the symbolic value.

Hi @frost-intel, what do you think?

Reply:
That's a much better solution. Thanks for pointing me towards `size_hints`.